server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java

```diff
@@ -142,8 +152,24 @@ String descriptCollectionStrategy() {
         return collectionStrategy.describe();
     }
 
+    public void setWeight(Weight weight) {
+        this.weight = weight;
+    }
+
     @Override
     public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        if (weight != null && weight.getQuery() instanceof MatchAllDocsQuery) {
+            if ((weight.count(ctx) == 0)
+                && Terms.getTerms(ctx.reader(), String.valueOf(((WithOrdinals.FieldData) valuesSource).indexFieldData.getFieldName()))
+                    .size() == 0) {
+                return LeafBucketCollector.NO_OP_COLLECTOR;
+                // } else if (weight.count(ctx) == ctx.reader().maxDoc() && weight.getQuery() instanceof MatchAllDocsQuery) {
+                // no deleted documents & top level query matches everything
+                // iterate over the terms - doc frequency for each termsEnum directly
+                // return appropriate LeafCollector
+            }
+        }
```

Comment:
You shouldn't need to explicitly check for `MatchAllDocsQuery`.

The interesting question is whether the query is "functionally" a match-all or match-none against the current segment.

Essentially, if `weight.count(ctx) == 0`, it's functionally a match none (and therefore will collect nothing from this segment). If `weight.count(ctx) == ctx.reader().maxDoc()`, it matches every document on the segment (and there are no deletes) so you can count the terms directly.

Reply:
@msfroh Now this case `weight.count(ctx) == 0` is what I was expecting when there were no terms in my documents. Interestingly, `my weight.count(ctx)` was non-zero in this case. 

My test-case was that I created an index with 2 term fields but my documents had only one of the terms. When trying to aggregate over the term for which was missing from my documents, I was expecting the `weight#count` to return 0.

Reply:
You need to distinguish between the query that produces the candidate documents for aggregation (the top-level query) and the field that you're aggregating over. This `weight` should be for the top-level query.

The match-none scenario is mildly interesting, with some possible impact on cases where we can skip large segments. (Small segments tend to be cheap, so skipping them isn't as impactful.)

Where things get more exciting is when the top-level query matches everything (either because it is a match-all query or because it happens to match everything). Then the `GlobalOrdinalsStringTermsAggregator` should be able avoid collecting anything, opting instead to populate buckets with the doc frequencies for each term.